# **Numpy (second) First Steps**

### A.J. Zerouali (21/06/28)

My first steps in Numpy. My objective is to get to do regular linear algebra in Python, including but not limited to: displaying a fixed number of decimals in Python,  matrix/tensor multiplications and inversion (in the mathematical sense), transposition, computing determinants and traces, finding eigenvalues and eigenvectors, diagonalization etc..

I'll start with Pierian Data's lectures in Section 5 and branch-out from there, because I need to be able to do in Python what I can do in Matlab. I will also look at Lazy Programmer's "Numpy Stack" (the free course).

___________________________________________

## **Contents**

I - Numpy arrays

II - Basic Linear Algebra with Numpy

In [1]:
# We all have to start somewhere...
import numpy as np

__________________________________________

# **I - Numpy arrays**

### Notes:

* It seems that Numpy works with row vectors instead of column vectors.

## **(1) From lists to arrays**

In [ ]:
# Here's how one creates arrays from lists in Python

In [2]:
mat_1_lst = [[1, 0 ,0], [0, 1 ,0], [0, 0 ,1]]
mat_1 = np.array(mat_1_lst)

In [3]:
mat_1

array([[1, 0, 0],
       [0, 1, 0],
       [0, 0, 1]])

In [53]:
# The dimensions of a tensor are given the SHAPE ATTRIBUTE:
print(f"mat_1.shape = {mat_1.shape}")

mat_1.shape = (3, 3)


In [54]:
# There's also the "dtype" attribute:
print(f"mat_1.dtype = {mat_1.dtype}")

mat_1.dtype = int32


**Note:** The .append() method is undefined for arrays. To concatenate arrays, one has to specify the axis of concatenation: **axis=0** means concatenate as new rows, **axis=1** means concatenate as new column.

In [55]:
vect_3x1 = np.random.rand(3).reshape(3,1)
print(f"vect_3x1 = {vect_3x1}")

vect_3x1 = [[0.87489546]
 [0.05154887]
 [0.58501145]]


In [67]:
e_1 = np.array([1.0, 0.0, 0.0]).reshape(3,1)
print(f"np.concatenate((vect_3x1, e_1)) = \n{np.concatenate((vect_3x1, e_1))}")
print(f"np.concatenate((vect_3x1, e_1),axis=0) = \n{np.concatenate((vect_3x1, e_1),axis=0)}")
print(f"np.concatenate((vect_3x1, e_1),axis=1) = \n{np.concatenate((vect_3x1, e_1),axis=1)}")

np.concatenate((vect_3x1, e_1)) = 
[[0.87489546]
 [0.05154887]
 [0.58501145]
 [1.        ]
 [0.        ]
 [0.        ]]
np.concatenate((vect_3x1, e_1),axis=0) = 
[[0.87489546]
 [0.05154887]
 [0.58501145]
 [1.        ]
 [0.        ]
 [0.        ]]
np.concatenate((vect_3x1, e_1),axis=1) = 
[[0.87489546 1.        ]
 [0.05154887 0.        ]
 [0.58501145 0.        ]]


## **(2) Some Numpy array generators**



In [23]:
# Identity matrix of size nxn
print(f"np.eye(3) = {np.eye(3)}")
print(f"np.eye(5) = {np.eye(5)}")

np.eye(3) = [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
np.eye(5) = [[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]]


In [17]:
# Makes matrices containing only zeroes and ones.
## Note that the 
mat_zeroes = np.zeros(5)
print(mat_zeroes)
print(f"mat_zeroes={mat_zeroes}")
print(f"type(mat_zeroes)={type(mat_zeroes)}")
mat_ones = np.ones((3,3))
print(f"mat_ones={mat_ones}")
print(f"type(mat_ones)={type(mat_ones)}")

# Note that when we write (n,m), n= # of rows and m= # of columns
mat_ones_23 = np.ones((2,3))
print(f"mat_ones_23={mat_ones_23}")
print(f"type(mat_ones_23)={type(mat_ones_23)}")

[0. 0. 0. 0. 0.]
mat_zeroes=[0. 0. 0. 0. 0.]
type(mat_zeroes)=<class 'numpy.ndarray'>
mat_ones=[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]
type(mat_ones)=<class 'numpy.ndarray'>
mat_ones_23=[[1. 1. 1.]
 [1. 1. 1.]]
type(mat_ones_23)=<class 'numpy.ndarray'>


In [5]:
# The Numpy arange() function, analogue of range()
## Why? Aren't they the same? Seems like the class of obj is different...
help(np.arange)

Help on built-in function arange in module numpy:

arange(...)
    arange([start,] stop[, step,], dtype=None)
    
    Return evenly spaced values within a given interval.
    
    Values are generated within the half-open interval ``[start, stop)``
    (in other words, the interval including `start` but excluding `stop`).
    For integer arguments the function is equivalent to the Python built-in
    `range` function, but returns an ndarray rather than a list.
    
    When using a non-integer step, such as 0.1, the results will often not
    be consistent.  It is better to use `numpy.linspace` for these cases.
    
    Parameters
    ----------
    start : number, optional
        Start of interval.  The interval includes this value.  The default
        start value is 0.
    stop : number
        End of interval.  The interval does not include this value, except
        in some cases where `step` is not an integer and floating point
        round-off affects the length of `out`.
   

In [15]:
arange_tst = np.arange(0,11,2)
print(f"np.arange(0,11,2) = arange_tst= {arange_tst}")
print(print(f"type(arange_tst) = {type(arange_tst)}"))

np.arange(0,11,2) = arange_tst= [ 0  2  4  6  8 10]
type(arange_tst) = <class 'numpy.ndarray'>
None


In [22]:
# np.linspace(a,b,N)
# This function return an array of N equally-spaced values in [a,b], including endpoints
print(f"np.linspace(0,0.75,5) = {np.linspace(0,0.75,5)}")
print(f"np.linspace(0,2,10+1) = {np.linspace(0,2,10+1)}")

np.linspace(0,0.75,5) = [0.     0.1875 0.375  0.5625 0.75  ]
np.linspace(0,2,10+1) = [0.  0.2 0.4 0.6 0.8 1.  1.2 1.4 1.6 1.8 2. ]


## **(3) Random Numpy array generators**



In [26]:
# The function np.random.rand(n,m) outputs an (n,m) with entries uniformly distributed in [0,1]
print(f"np.random.rand(4)) = {np.random.rand(4)}")
print(f"np.random.rand(3,2)) = {np.random.rand(3,2)}")


np.random.rand(4)) = [0.61132354 0.78788057 0.7107116  0.36090009]
np.random.rand(3,2)) = [[7.52197668e-01 3.92801681e-01]
 [7.86605547e-01 5.72392202e-01]
 [2.88027670e-05 3.09086962e-01]]


In [28]:
# The function np.random.randn(n,m) outputs an (n,m) with entries random entries following N(0,1)
print(f"np.random.randn(4)) = {np.random.randn(4)}")
print(f"np.random.randn(3,2)) = {np.random.randn(3,2)}")
print(f"np.random.randn(4,4)) = {np.random.randn(4,4)}")

np.random.randn(4)) = [ 0.26757865  0.67275968  0.51562024 -0.83308285]
np.random.randn(3,2)) = [[ 0.11951656 -0.14578893]
 [ 0.11266493 -1.09624761]
 [ 0.04495483 -0.83181161]]
np.random.randn(4,4)) = [[ 0.19975258 -0.62643152 -0.47108995 -1.60135334]
 [-0.3842183   0.6014606  -0.53570325  0.55610201]
 [-0.18844925 -1.49137671  0.89171474  0.29094652]
 [-2.09739588 -1.55795607 -1.87765275 -1.22482773]]


In [34]:
# The function np.random.randint(a,b,(n,m)) returns an (n,m) array of integers randomly drawn in [a,b[ (uniformly)
# If (n,m) isn't specified it returns ONE random integer. If replaced by an integer N, returns a (1,N) array
help(np.random.randint)

Help on built-in function randint:

randint(...) method of numpy.random.mtrand.RandomState instance
    randint(low, high=None, size=None, dtype=int)
    
    Return random integers from `low` (inclusive) to `high` (exclusive).
    
    Return random integers from the "discrete uniform" distribution of
    the specified dtype in the "half-open" interval [`low`, `high`). If
    `high` is None (the default), then results are from [0, `low`).
    
    .. note::
        New code should use the ``integers`` method of a ``default_rng()``
        instance instead; please see the :ref:`random-quick-start`.
    
    Parameters
    ----------
    low : int or array-like of ints
        Lowest (signed) integers to be drawn from the distribution (unless
        ``high=None``, in which case this parameter is one above the
        *highest* such integer).
    high : int or array-like of ints, optional
        If provided, one above the largest (signed) integer to be drawn
        from the distributi

In [33]:
print(f"np.random.randint(0,10,(3,3)) = {np.random.randint(0,10,(3,3))}")
print(f"np.random.randint(0,101) = {np.random.randint(0,101)}")
print(f"np.random.randint(0,101,10) = {np.random.randint(0,101,10)}")

np.random.randint(0,10,(3,3)) = [[2 3 7]
 [3 6 1]
 [9 5 6]]
np.random.randint(0,101) = 25
np.random.randint(0,101,10) = [ 75  42 100  76  63  57  85   6  19  54]


## **(4) Basic attributes and methods of arrays**



* **Reshape** (What use is this?)


In [42]:
mat_1x25 = np.random.randn(25)
mat_5x5 = mat_1x25.reshape(5,5)
print(f"Using mat_5x5 = mat_1x25.reshape(5,5), the matrix:\n"+\
      f"mat_1x25 = {mat_1x25}\ngets reshaped into:\n mat_5x5 = {mat_5x5}")

Using mat_5x5 = mat_1x25.reshape(5,5), the matrix:
mat_1x25 = [ 1.74925262  0.47830289 -0.78028406 -1.57692129 -1.3587184  -0.58549998
  0.66175843  0.59609693  0.56983428  0.371461   -1.57965455 -0.22493176
  1.30791295  0.92605807 -1.88707958 -0.55266442  0.02120173 -0.43520047
  0.79052935 -1.2875192  -0.95478357  0.58732485 -1.6012867   0.47517854
 -0.68023292]
gets reshaped into:
 mat_5x5 = [[ 1.74925262  0.47830289 -0.78028406 -1.57692129 -1.3587184 ]
 [-0.58549998  0.66175843  0.59609693  0.56983428  0.371461  ]
 [-1.57965455 -0.22493176  1.30791295  0.92605807 -1.88707958]
 [-0.55266442  0.02120173 -0.43520047  0.79052935 -1.2875192 ]
 [-0.95478357  0.58732485 -1.6012867   0.47517854 -0.68023292]]


* **Min and max values in an array, argmin and argmax for their indices**

**Note:** Argmin and argmax don't return an index (i,j), they return the index i of min/max for the *flattened* array... How do you look (i,j) index?


In [44]:
matrix_4x4 = np.random.randn(4,4)
print(f"matrix_4x4 = {matrix_4x4}")

matrix_4x4 = [[-0.5650311  -0.26414763 -0.03989725 -0.53350167]
 [-0.43921842  0.77420045  0.3203596   0.4898972 ]
 [-1.15909817 -1.24049477  0.45167297 -0.62922376]
 [ 0.22344286  0.92638329 -0.33933331  0.43098806]]


In [45]:
print(f"matrix_4x4.min() = {matrix_4x4.min()}")
print(f"matrix_4x4.max() = {matrix_4x4.max()}")
print(f"matrix_4x4.argmin() = {matrix_4x4.argmin()}")
print(f"matrix_4x4.argmax() = {matrix_4x4.argmax()}")

matrix_4x4.min() = -1.2404947733249236
matrix_4x4.max() = 0.9263832860947631
matrix_4x4.argmin() = 9
matrix_4x4.argmax() = 13


## **(5) Statistical properties of arrays**

This is from Lecture 6 in Section 2 of Lazy Programmer's "Numpy stack" course.

Below, we start with a random array with a large number of entries sampled from $\mathcal{N}(0,1)$. The mean, the variance, and the standard deviation can be obtained from Numpy arrays' methods **.mean()**, **.var()** and **.std()**.



In [2]:
N = 100000
randn_array = np.random.randn(N)

In [6]:
print(f"Mean: randn_array.mean() = {randn_array.mean()}\n \
        Variance: randn_array.var() = {randn_array.var()}\n \
        Std. dev.: randn_array.std() = {randn_array.std()}")

Mean: randn_array.mean() = -0.0018286991126917067
         Variance: randn_array.var() = 0.9977995181854865
         Std. dev.: randn_array.std() = 0.9988991531608616


In [7]:
randn_array.std() **2

0.9977995181854865

Next, we look at these statistical properties for matrices. In practice, matrices will be used as CSV tables, i.e. each column will correspond to a feature while each row will correspond to a sample.

To compute the mean of each column, use **.mean(axis = 0)**, and **.mean(axis = 1)** to compute the mean of each column. This works similarly for variance and std. dev.

The confusing point here is that normally, axis 0 would represent rows while axis 1 would represent columns. The command array_variable.mean(axis = 0) means **compute means along axis 0 and preserve the number along axis 1 (# of columns)**. This is the same idea as for the concatenation of arrays above.

In [9]:
randn_table = np.random.randn(N,10)

In [13]:
# Mean of each column:
mean_col_table = randn_table.mean(axis = 0)
mean_col_table.shape

(10,)

In [14]:
# Mean of each row:
mean_row_table = randn_table.mean(axis = 1)
mean_row_table.shape

(100000,)

In [ ]:
print(f"randn_table.var(axis = 1) = {randn_table.var(axis = 1)}")
print(f"randn_table.std(axis = 0) = {randn_table.std(axis = 0)}")

Last but not least, covariance. This is given by the Numpy function **np.cov()** (not an attribute).

**Important notes:** 

* There's something wrong with this covariance function, because it doesn't agree with the mathematical definition of the covariance matrix of a random vector. 
* This is not the same convention as the remainder of the Numpy stack. See last third of relevant lecture.
* **I'll get back to this later...**

In [23]:
vect_1x3 = np.random.rand(3)
print(f"np.cov(vect_1x3) = {np.cov(vect_1x3)}")
print(f"np.cov(vect_1x3,vect_1x3) = {np.cov(vect_1x3,vect_1x3)}")
print(f"np.cov(vect_1x3,rowvar=False) = {np.cov(vect_1x3,rowvar=False)}")

np.cov(vect_1x3) = 0.015652865711285973
np.cov(vect_1x3,vect_1x3) = [[0.01565287 0.01565287]
 [0.01565287 0.01565287]]
np.cov(vect_1x3,rowvar=False) = 0.015652865711285973


## **(6) Additional notes:**

Here are various notes, comments and tricks.

* **Slicing of arrays, broadcasting and modifications:** It looks like once an array is created, broadcasting Numpy treats every slicing of it as an array of pointers, see Lect. 20 in Sec. 5 of Portilla's course. If a slice is modified then so is the original array.
* **Arrays and comparison statements:** See Lect. 20 in Sec. 5 of Portilla's DSML course. Using an array with a comparison statement returns an arrays of Booleans. This is useful for conditional selection. Relevant because apparently used a lot with Pandas.



In [39]:
array = np.arange(10)
bool_array = array > 5

print(f"Have:\n array = {array},\nand then:\nbool_array= (array > 5) = {bool_array}\n\
Then: array[bool_array] = {array[bool_array]}.\n\
More shortly: array[array <8] = {array[array <8]}.")

Have:
 array = [0 1 2 3 4 5 6 7 8 9],
and then:
bool_array= (array > 5) = [False False False False False False  True  True  True  True]
Then: array[bool_array] = [6 7 8 9].
More shortly: array[array <8] = [0 1 2 3 4 5 6 7].


* **Functions of arrays:** When applying a function to an array, Numpy executes the function element-wise. For instance, 

In [48]:
pi = np.pi
from numpy import cos
from numpy import sin
from numpy import exp
t = pi/3
X = np.array([[0, -t], [t, 0]])

In [50]:
print(f"X = {X}")
print(f"exp(X) = {exp(X)}")
print(f"X**2 +1 = {X**2 +1}")

X = [[ 0.         -1.04719755]
 [ 1.04719755  0.        ]]
exp(X) = [[1.         0.35091981]
 [2.84965391 1.        ]]
X**2 +1 = [[1.         2.09662271]
 [2.09662271 1.        ]]


* **Universal array functions:** (Lect 21 of Portilla's DSML, Sec 5) Here is a useful link, with a list of universal functions available in Numpy and that can be applied to arrays:
https://numpy.org/doc/stable/reference/ufuncs.html

____________________________

# **II - Basic Linear Algebra with Numpy**

### Notes:

* This is from Lazy Programmer's Numpy Stack course, Section 2.
* Need to say something about broadcasting
* Need to say something about functions of matrices

## **(1) Addition and broadcasting**

In [68]:
vect_4x1 = np.random.randint(0,11,(4,1))
print(f"vect_4x1 = {vect_4x1}")
e_1= np.array([1.0, 0.0, 0.0, 0.0,]).reshape(4,1)
print(f"e_1 = {e_1}")

vect_4x1 = [[2]
 [6]
 [4]
 [4]]
e_1 = [[1.]
 [0.]
 [0.]
 [0.]]


In [69]:
# Usual matrix addition is done as one would expect
vect_4x1+e_1

array([[3.],
       [6.],
       [4.],
       [4.]])

In [70]:
# If we just add a (1,1) array to the vector, Numpy broadcasts it as follows:
vect_4x1+np.array([1])
# Add a vector with size different from (1,1) or (4,1) is undefined however

array([[3],
       [7],
       [5],
       [5]])

## **(2) Scalar multiplication**

Exactly as one would expect.

In [71]:
Iden_3 = np.eye(3)

In [72]:
print(f"10*Iden_3 = {10*Iden_3}")
print(f"Iden_3*10 = {Iden_3*10}")

10*Iden_3 = [[10.  0.  0.]
 [ 0. 10.  0.]
 [ 0.  0. 10.]]
Iden_3*10 = [[10.  0.  0.]
 [ 0. 10.  0.]
 [ 0.  0. 10.]]


In [73]:
print(f" 5*vect_4x1 = {5*vect_4x1} ")

 5*vect_4x1 = [[10]
 [30]
 [20]
 [20]] 


## **(3) The dot product of vectors**

Because of the heavy broadcasting in Numpy, the usual "\*" operator just multiplies component-wise:

In [88]:
v = np.array([1,2,3,4]).reshape(4,1)
w = np.array([5,6,7,8]).reshape(4,1)

In [89]:
v

array([[1],
       [2],
       [3],
       [4]])

In [102]:
w

array([[5],
       [6],
       [7],
       [8]])

In [95]:
v+w

array([[ 6],
       [ 8],
       [10],
       [12]])

In [90]:
print(f"v*w = {v*w}")

v*w = [[ 5]
 [12]
 [21]
 [32]]


One way of composing the dot product is then:

In [93]:
v_dot_w_1 = (v*w).sum()
v_dot_w_1

70

The Numpy function for the dot product is **np.dot(v,w)**, but beware **seems to only work on row vectors**:

- If we compile np.dot(v,w), Numpy gives an error message.

- v and w must be reshaped.

In [104]:
np.dot(v.reshape(4), w.reshape(4))

70

In [103]:
np.dot(np.array([1,2,3,4]),np.array([5,6,7,8]))

70

The right way to get the dot product of column vectors is to do the following (see matrices below for the matrix product):

In [131]:
print(f"np.dot(v.T,w) = {np.dot(v.T,w)}")

np.dot(v.T,w) = [[70]]


... nope, I don't like this. Other ways of computing the dot product:

In [107]:
a = v.reshape(4)
b= w.reshape(4)
# Here's another way to do the dot product
a @ b

70

In [110]:
# ... and another way:
a.dot(b)

70

For the geometric interpretation of the dot product: 

$$ \vec{u}\bullet\vec{v} = |\!|\vec{u}|\!|\cdot |\!|\vec{v}|\!|\cdot \cos(\theta_{v,w}), $$

there's a dedicated function for the norm in a subpackage of Numpy, the "linalg" package:

In [111]:
norm_v = np.linalg.norm(v)

In [112]:
norm_v

5.477225575051661

In [117]:
norm_w = np.linalg.norm(w)
theta_v_w = np.arccos( (v*w).sum()/(norm_v*norm_w ) )
print(f"theta_v_w = {theta_v_w}")

theta_v_w = 0.25019592042251176


## **(4) Matrices**



### **4.a - Indices**: 

Aside from the list indices, arrays in numpy also allow the use of matrix notation **M[i,j]** for indices.

In [120]:
mat_2x3 = np.array([[1, 2, 3], [4, 5, 6]])
print(f"mat_2x3 = {mat_2x3}")
print(f"List indexing: mat_2x3[0][1] = {mat_2x3[0][1]}")
print(f"Matrix indexing: mat_2x3[1,1] = {mat_2x3[1,1]}")

mat_2x3 = [[1 2 3]
 [4 5 6]]
List indexing: mat_2x3[0][1] = 2
Matrix indexing: mat_2x3[1,1] = 5


The index slicing used for lists is also valid here.

### **4.b - Transposes:** 
This is an important command, done using **array_var.T**.

In [74]:
print(f"vect_4x1.T = {vect_4x1.T}")

vect_4x1.T = [[2 6 4 4]]


In [77]:
Mat_2x3  = np.random.randint(0,10+1,(2,3))
print(f"Mat_2x3 = {Mat_2x3}")
print(f"Mat_2x3.T = {Mat_2x3.T}")

Mat_2x3 = [[6 2 6]
 [9 8 4]]
Mat_2x3.T = [[6 9]
 [2 8]
 [6 4]]


### **4.c - Rows/Columns:** 
One can also extract rows and columns from a given matrix using **M[:,j]** and **M[i,:]**, a little like in Matlab.

In [123]:
# Note the return to row vectors for the first case
print(f"Third column of mat_2x3: mat_2x3[:,2] = {mat_2x3[:,2]}")
print(f"Third column of mat_2x3: mat_2x3[:,2].T = {(mat_2x3[:,2]).T}")
print(f"First row of mat_2x3: mat_2x3[0,:] = {mat_2x3[0,:]}")

Third column of mat_2x3: mat_2x3[:,2] = [3 6]
Third column of mat_2x3: mat_2x3[:,2].T = [3 6]
First row of mat_2x3: mat_2x3[0,:] = [1 2 3]


### **4.d - Matrix multiplication:** 

Next, **matrix multiplication** is done using the .dot() functions, or the "@" operator:

In [126]:
A = np.array([[0.5, 0], [0, 0.5]])
print(f"A = {A}")
B = np.array([[4, 6], [3, 8]])
print(f"B = {B}")

A = [[0.5 0. ]
 [0.  0.5]]
B = [[4 6]
 [3 8]]


In [132]:
print(f"A.dot(B) = {A.dot(B)}")
print(f"np.dot(A,B) = {np.dot(A,B)}")
print(f"A @ B = {A @ B}")

A.dot(B) = [[2.  3. ]
 [1.5 4. ]]
np.dot(A,B) = [[2.  3. ]
 [1.5 4. ]]
A @ B = [[2.  3. ]
 [1.5 4. ]]


### **4.e - Matrix Inversion:** 

Inversion calls upon the linalg package too: we use **np.linalg.inv()**

In [137]:
pi = np.pi
from numpy import cos
from numpy import sin

In [140]:
orth_mat = np.array([[cos(pi/3), -sin(pi/3)], [sin(pi/3), cos(pi/3)]])
print(f"orth_mat = {orth_mat}")
print(f"orth_mat.T = {orth_mat.T}")
print(f"np.linalg.inv(orth_mat) = {np.linalg.inv(orth_mat)}")

orth_mat = [[ 0.5       -0.8660254]
 [ 0.8660254  0.5      ]]
orth_mat.T = [[ 0.5        0.8660254]
 [-0.8660254  0.5      ]]
np.linalg.inv(orth_mat) = [[ 0.5        0.8660254]
 [-0.8660254  0.5      ]]


### **4.e - Trace and Determinant:** 

The trace function is in Numpy, but the determinant function is in linalg. They're respectively called using **np.trace()** and **np.linalg.det()**

In [142]:
A = 2*orth_mat
print(f"A = 2*orth_mat = {2*orth_mat}")
print(f"tr(A) = np.trace(A) = {np.trace(A)}")
print(f"det(A) = np.linalg.det(A) = {np.linalg.det(A)}")

A = 2*orth_mat = [[ 1.         -1.73205081]
 [ 1.73205081  1.        ]]
tr(A) = np.trace(A) = 2.0000000000000004
det(A) = np.linalg.det(A) = 4.000000000000001


**Note:** The round-off errors are pretty obvious in Python:

In [145]:
print(f"np.dot(orth_mat,np.linalg.inv(orth_mat))\
= {np.dot(orth_mat,np.linalg.inv(orth_mat))}")

np.dot(orth_mat,np.linalg.inv(orth_mat))= [[1.00000000e+00 1.48741681e-17]
 [7.03853194e-17 1.00000000e+00]]


### **4.g - Diagonal matrices and diagonal elements:** 
Getting the diagonal of a matrix is done using the **np.diag()** function

In [146]:
A = np.array([[1,2,3],[4,5,6],[7,8,9]])
print(f"A = {A}")
print(f"np.diag(A) = {np.diag(A)}")

A = [[1 2 3]
 [4 5 6]
 [7 8 9]]
np.diag(A) = [1 5 9]


As in Matlab, one can create an nxn diagonal matrix using **np.diag()** also:

In [147]:
B = np.diag([-1, 2, 1])
print(f"B = np.diag([-1, 2, 1]) = {B}")

B = np.diag([-1, 2, 1]) = [[-1  0  0]
 [ 0  2  0]
 [ 0  0  1]]


## **(6) Eigenvalues and eigenvectors of matrices:**

These call upon the linalg package.

In [159]:
A = np.array([[1,2],[3,4]])
E_A = np.linalg.eig(A)
print(f"E_A = np.linalg.eig(A) = {E_A}")

E_A = np.linalg.eig(A) = (array([-0.37228132,  5.37228132]), array([[-0.82456484, -0.41597356],
       [ 0.56576746, -0.90937671]]))


In the above example, $\lambda_+ \simeq 5.37$ and $\lambda_- \simeq -0.37$. The output of linalg.eig() is a tuple, whose first entry is a **vector containing eigenvalues**, with second entry a **matrix whose columns are normalized eigenvectors**. Note that the output is a tuple.

In [161]:
eig_v_neg = E_A[1][:,0]
eig_v_pos = E_A[1][:,1]

print(f"A.eig_v_neg = {A @eig_v_neg}")
print(f"A.eig_v_pos = {A @eig_v_pos}")

A.eig_v_neg = [ 0.30697009 -0.21062466]
A.eig_v_pos = [-2.23472698 -4.88542751]


In [163]:
# The ultimate test:
P = E_A[1]
D = (np.linalg.inv(P)) @ (A @ P)
print(f"D = (P^-1).A.P = {D}")

D = (P^-1).A.P = [[-3.72281323e-01  6.87203782e-16]
 [-4.89517862e-17  5.37228132e+00]]


For complex eigen-elements:

In [148]:
eig_orth_mat = np.linalg.eig(orth_mat)
print(f"np.linalg.eig(orth_mat) = {eig_orth_mat}")

np.linalg.eig(orth_mat) = (array([0.5+0.8660254j, 0.5-0.8660254j]), array([[0.70710678+0.j        , 0.70710678-0.j        ],
       [0.        -0.70710678j, 0.        +0.70710678j]]))


Notice that $j = \sqrt{-1}$ for Numpy. To isolate it use **1j**:

In [168]:
(1j)**2

(-1+0j)

In [169]:
(2*1j)**2

(-4+0j)

## **(7) On numerical precision:**

Lazy Programmer makes a note on numerical precision in Python's linear algebra computations. Returning to the diagonalization of our last example of A:

In [170]:
# The matrix
A = np.array([[1,2],[3,4]])
E_A = np.linalg.eig(A)

# Eigenvalues
lambda_A = E_A[0]

# Eigenvectors
eig_v_neg = E_A[1][:,0]
eig_v_pos = E_A[1][:,1]
P = E_A[1]

# Diagonal computed numerically
D = (np.linalg.inv(P)) @ (A @ P)

D_1 = np.diag(lambda_A)

The diagonalization of A doesn't return the eigenvalues that Numpy computed:

In [171]:
D == D_1

array([[ True, False],
       [False, False]])

The correct way of comparing such matrices is to use **np.allclose()**:

In [172]:
np.allclose(D, D_1)

True

## **(8) Solving linear systems:**

There's an optimized function to solve linear systems with Numpy, which is far better than using the standard $A^{-1}\cdot\vec{b}$.

In [176]:
A = np.array([[1,2],[3,4]])
b = np.array([[10.0],[11.0]])
x_1 = np.array([[-9.0],[19/2]])
print(f"A.x_1 = {np.dot(A,x_1)}")

A.x = [[10.]
 [11.]]


In [178]:
x = np.linalg.solve(A,b)
print(f"np.linalg.solve(A,b) = {x}")

np.linalg.solve(A,b) = [[-9. ]
 [ 9.5]]


**Remarks:** Add examples of systems with $\det A = 0$ and $\vec{b}\not\in \mathrm{Ran}(A)$

## **(9) Comparing execution times (Exercise)**

This is an exercise in Lazy Programmer's Numpy stack course (Lect. 10, Sec. 02). The question is to compare the execution times of ab-initio matrix multiplication (using lists and loops) with Numpy's function. The general approach to this experimant is explained in Lecture 6.

Here the Colab Notebook: https://colab.research.google.com/drive/1VdLyfnop7Tt4wx4BmS6WHSSQg3n5wXrX

In [2]:
# Generate 2 random 3x3 matrices with real entries in [0,100]
A = 100*np.random.rand(3,3)
B = 100*np.random.rand(3,3)

In [3]:
A.shape

(3, 3)

In [4]:
print(f"A = {A}")
print(f"B = {B}")

A = [[83.32511657 51.59176918  2.85423485]
 [ 5.72925842 80.43847626 57.72360123]
 [25.1357724  77.72802812 29.74011825]]
B = [[45.42243174 18.38255011 15.36310348]
 [42.62974536  8.59799226 15.77128572]
 [51.04616814 10.1725631  40.10964082]]


In [10]:
n = 2-1; m= 3-1
print(f"A[{n+1},{m+1}] = {A[n,m]}")
print(f"A[{m+1},{n+1}] = {A[m,n]}")

A[2,3] = 57.72360122528549
A[3,2] = 77.72802812006657


In [12]:
C = np.dot(A,B)
D = np.dot(B,A)
print(f"C = A.B = {C}")
print(f"D = B.A = {D}")

C = A.B = [[6129.87115412 2004.34864631 2208.283255  ]
 [6635.87726267 1384.12475179 3671.91029324]
 [5973.37302829 1432.89780894 2804.89987263]]
D = B.A = [[4276.31127147 5016.23167305 1647.65379352]
 [3997.81206864 4116.82431818 1087.02228332]
 [5319.89595577 6569.47088916 1925.76018888]]


A basic function to compute matrix products. If $C = A\cdot B$ with $A = [a_{ij}]$ and $B = [b_{ij}]$, then:

$c_{ij} = \sum_{k=1}^{n}a_{i,k}\cdot b_{k,j}$.


In [37]:
# My multiplication function for square matrices of same size
def Mult(A,B):
    # Just for a test, don't add too much
    n = A.shape[0]
    A_dot_B = np.zeros((n,n))
    for i in range(n):
        for j in range(n):
            for k in range(n):
                A_dot_B[i,j] += A[i,k]*B[k,j]
    return A_dot_B
        

In [38]:
C_1 = Mult(A,B)

In [42]:
print(f"Ab-initio multiplication: Mult(A,B) = \n{C_1}")
print(f"Numpy multiplication: np.dot(A,B) = \n{C}")
print(f"How close: np.allclose(C,C_1) = {np.allclose(C,C_1)}")

Ab-initio multiplication: Mult(A,B) = 
[[6129.87115412 2004.34864631 2208.283255  ]
 [6635.87726267 1384.12475179 3671.91029324]
 [5973.37302829 1432.89780894 2804.89987263]]
Numpy multiplication: np.dot(A,B) = 
[[6129.87115412 2004.34864631 2208.283255  ]
 [6635.87726267 1384.12475179 3671.91029324]
 [5973.37302829 1432.89780894 2804.89987263]]
How close: np.allclose(C,C_1) = True


Next, import a module to time the execution time of Mult() and of np.dot().

I'll use datetime.now(), whose output is a tuple (Yr, Mo, Day, Hr, min, sec, microseconds).

(I will not use timeit for now)

In [46]:
from datetime import datetime

In [60]:
datetime.now()

datetime.datetime(2021, 6, 29, 13, 1, 20, 108385)

In [70]:
t_0 = datetime.now()
A = 100*np.random.rand(3,3)
B = 100*np.random.rand(3,3)
C_1 = Mult(A,B)
t_1 = datetime.now()

print(f"Execution time for Mult(A,B) took: {1000000*(t_1-t_0)} microsec.")

Execution time for Mult(A,B) took: 0:00:00 microsec.


In [68]:
t_0

datetime.datetime(2021, 6, 29, 13, 6, 48, 631165)

In [69]:
t_1

datetime.datetime(2021, 6, 29, 13, 6, 48, 631165)

In [79]:
T = 1000000
t_0 = datetime.now()
for i in range(T):
    A = 100*np.random.rand(3,3)
    B = 100*np.random.rand(3,3)
    C_1 = Mult(A,B)
t_1 = datetime.now()

print(f"Execution of Mult(A,B) {T} times took: {(t_1-t_0)} sec.")

s_0 = datetime.now()
for i in range(T):
    A = 100*np.random.rand(3,3)
    B = 100*np.random.rand(3,3)
    C = np.dot(A,B)
s_1 = datetime.now()

print(f"Execution of np.dot(A,B) {T} times took: {(s_1-s_0)} sec.")

print(f"np.dot() is {(t_1-t_0)/(s_1-s_0)} faster than Mult()")


Execution of Mult(A,B) 1000000 times took: 0:00:42.959334 sec.
Execution of np.dot(A,B) 1000000 times took: 0:00:07.651654 sec.
np.dot() is 5.6143853342035595 faster than Mult()
